tests/contrib/kafka/test_kafka.py

```diff
@@ -666,38 +666,111 @@ def test_data_streams_default_context_propagation(dummy_tracer, consumer, produc
 
 
 # It is not currently expected for kafka produce and consume spans to connect in a trace
-def test_tracing_context_is_not_propagated(dummy_tracer, consumer, producer, kafka_topic):
+def test_tracing_context_is_not_propagated_by_default(dummy_tracer, consumer, producer, kafka_topic):
     Pin.override(producer, tracer=dummy_tracer)
     Pin.override(consumer, tracer=dummy_tracer)
 
-    test_string = "context test"
+    test_string = "context test no propagation"
+    test_key = "context test key no propagation"
     PAYLOAD = bytes(test_string, encoding="utf-8") if six.PY3 else bytes(test_string)
 
-    producer.produce(kafka_topic, PAYLOAD, key="test_key")
+    producer.produce(kafka_topic, PAYLOAD, key=test_key)
     producer.flush()
 
     message = None
     while message is None or str(message.value()) != str(PAYLOAD):
         message = consumer.poll(1.0)
 
     # message comes back with expected test string
-    assert message.value() == b"context test"
+    assert message.value() == b"context test no propagation"
 
+    consume_span = None
     traces = dummy_tracer.pop_traces()
     produce_span = traces[0][0]
-    consume_span1 = traces[1][0]
-    consume_span2 = traces[2][0]
+    for trace in traces:
+        for span in trace:
+            if span.get_tag("kafka.received_message") == "True":
+                if span.get_tag("kafka.message_key") == test_key:
+                    consume_span = span
 
     # kafka.produce span is created without a parent
     assert produce_span.name == "kafka.produce"
     assert produce_span.parent_id is None
 
     # None of the kafka.consume spans have parents
-    assert consume_span1.name == "kafka.consume"
-    assert consume_span1.parent_id is None
-    assert consume_span2.name == "kafka.consume"
-    assert consume_span2.parent_id is None
+    assert consume_span.name == "kafka.consume"
+    assert consume_span.parent_id is None
 
     # None of these spans are part of the same trace
-    assert produce_span.trace_id != consume_span1.trace_id
-    assert consume_span1.trace_id != consume_span2.trace_id
+    assert produce_span.trace_id != consume_span.trace_id
+
+
+# Propagation should work when enabled
+def test_tracing_context_is_propagated_when_enabled(ddtrace_run_python_code_in_subprocess):
+    code = """
```

Comment:
Using [this decorator](https://github.com/DataDog/dd-trace-py/blob/5cb0b45c7179c21f203bff2ebe4008a7b112d0a3/tests/contrib/loguru/test_loguru_logging.py#L69) means you won't have to put such a large amount of Python code into a string.

Reply:
I recommitted with test refactor with the decorator. However the tests are now failing due to connection errors from kafka being printed to stdout. @mabdinur recommended to just use the original approach and keep the code within a string so that I am able to control the assertions.

Reply:
This test is reliant on pytest fixtures and pytest fixtures can not be used with the `@pytest.mark.subprocess(..)`. This is something we can improve in our testing framework.  

This test is less than ideal but works for now. Maybe another approach could be to simplify this test to mock the producer and consumer. With this approach we no longer need the `consumer, producer, kafka_topic` fixtures.

Reply:
IMO mocking the producer / consumer is not ideal since we want to test the propagation is working correctly.